apps/system/js/utility_tray_motion.js

```diff
@@ -0,0 +1,409 @@
+ *
+ *------------------------------------------------------------------------------
+ * TRACKING SCROLL POSITION
+ *
+ * Because we now support asynchronous scrolling (APZ), we can never predict
+ * the element's exact scroll position at any given time. We can monitor touch
+ * and scroll events, but `scrollTop` is always an out-of-date snapshot.
+ * We cannot cancel scroll events synchronously, nor can we rely on touch events
+ * to calculate the element's curent scroll offset, because the platform applies
+ * momentum itself.
+ *
+ * Fortunately, we can compensate for this lack of control by tracking the
+ * element's scrollTop whenever a touch or scroll event occurs. In this way,
+ * we can update the tray state ('open', 'closed', 'opening', 'closing') in
+ * response to changes in scroll position. When we need to move the tray
+ * manually, we use `scrollTo({ behavior: 'smooth'})`.
+ *
+ * @param {Element} el
+ */
+function UtilityTrayMotion(el) {
+  this.el = el;
+
+  ['touchstart', 'touchmove', 'touchend', 'touchcancel'].forEach((type) => {
+    this.el.addEventListener(type, this._ontouch.bind(this));
+  });
+  this.el.addEventListener('scroll', this._onscroll.bind(this));
+
+  // Since scroll events are unpredictable, we can't reliably calculate scroll
+  // velocity without averaging over a few samples.
+  const VELOCITY_WINDOW_MS = 400;
+  this._velocityAverager = new TimeWindowedAverager(VELOCITY_WINDOW_MS);
+
+  // We track the most recent time since the last touch event, in order to
+  // prevent misinterpreted touches.
+  this._lastTouchTime = 0;
+
+  this.isTouching = false;
+  this.isRecoiling = false;
+  this.currentTouch = null; // used by UtilityTray for event forwarding
+
+  // Begin closed. Note that scrollTop may be the inverse of what you'd expect;
+  // a scrollTop of zero indicates the tray is open; a scrollTop of scrollTopMax
+  // indicates the tray is closed.
+  this._setState('closed');
+  this.position = this.el.scrollTop = this.el.scrollTopMax;
+
+  // We may need to reposition ourselves when the visible buttons change.
+  // (CSS properly adjusts width and height, but the tray may be stuck in the
+  // middle of the screen.)
+  window.addEventListener('software-button-enabled',
+    this.requestFuturePosition.bind(this));
+  window.addEventListener('software-button-disabled',
+    this.requestFuturePosition.bind(this));
+}
+
+// Avoid interfering with the scroll position if we have recently touched the
+// element. We don't have enough information at touchend to determine velocity
+// accurately, which could lead to incorrect scroll snapping.
+const LAST_TOUCH_MIN_DELAY_MS = 200;
+
+// We don't know that the element's scroll position has completely stopped until
+// we see the scroll position not moving after a bit of time. Whenever we think
+// the scroll position is changing, we set a timer to check the position again
+// in the near future.
+const FUTURE_POSITION_DELAY_MS = 250;
```

Comment:
Let's add a reference to https://bugzilla.mozilla.org/show_bug.cgi?id=1172171 here :)